proxy/proxy.go

```diff
@@ -975,32 +976,27 @@ func (p *Proxy) do(ctx *context) error {
 			p.metrics.IncErrorsBackend(ctx.route.Id)
 
 			if perr.DialError() && ctx.route.IsLoadBalanced {
-				// here we do a transparent retry, because we know it's safe to do
-				origRoute := ctx.route.Me
-				if ctx.route.Next != nil && origRoute != ctx.route.Next {
-					ctx.route = ctx.route.Next
-				} else if ctx.route.Head != nil && origRoute != ctx.route.Head {
-					ctx.route = ctx.route.Head
-				}
-
 				if ctx.proxySpan != nil {
 					ctx.proxySpan.Finish()
 					ctx.proxySpan = nil
 				}
 
 				tracing.LogKV("retry", ctx.route.Id, ctx.Request().Context())
 
+				// lbalgorithm has to be the last route
```

Comment:
this limitation is a bit of a leaky abstraction. There could be an interface:

```
type LBAlgorithmFilter interface { ApplyLB(filters.FilterContext) }
```

and we could find the last filter from the end of the filter chain that implements this interface, and explicitly call it.

Reply:
If we can do this before the proxy.go, I am fine.